tmt/package_managers/__init__.py

```diff
@@ -0,0 +1,179 @@
+
+    Decorate a package manager plugin class to register a package manager.
+    """
+
+    def _provides_package_manager(package_manager_cls: PackageManagerClass) -> PackageManagerClass:
+        _PACKAGE_MANAGER_PLUGIN_REGISTRY.register_plugin(
+            plugin_id=package_manager,
+            plugin=package_manager_cls,
+            logger=tmt.log.Logger.get_bootstrap_logger())
+
+        return package_manager_cls
+
+    return _provides_package_manager
+
+
+def find_package_manager(name: 'GuestPackageManager') -> 'PackageManagerClass':
+    """
+    Find a package manager by its name.
+
+    :raises GeneralError: when the plugin does not exist.
+    """
+
+    plugin = _PACKAGE_MANAGER_PLUGIN_REGISTRY.get_plugin(name)
+
+    if plugin is None:
+        raise tmt.utils.GeneralError(
+            f"Package manager '{name}' was not found in package manager registry.")
+
+    return plugin
+
+
+def escape_packages(*packages: Installable) -> Iterator[str]:
+    for package in packages:
+        if isinstance(package, Path):
+            yield str(package)
+
+        else:
+            yield shlex.quote(package)
+
+
+# TODO: find a better name, "options" is soooo overloaded...
+@dataclasses.dataclass(frozen=True)
+class Options:
+    #: A list of packages to exclude from installation.
+    excluded_packages: list[Package] = dataclasses.field(default_factory=list)
+
+    #: If set, a failure to install a given package would not cause an error.
+    skip_missing: bool = False
+
+    #: If set, check whether the package is already installed, and do not
+    #: attempt to install it if it is already present.
+    check_first: bool = True
+
+    #: If set, install packages under this path instead of the usual system
+    #: root.
+    install_root: Optional[Path] = None
+
+    #: If set, instruct package manager to behave as if the distribution release
+    #: was ``release_version``.
+    release_version: Optional[str] = None
+
+
+class PackageManager(tmt.utils.Common):
+    """ A base class for package manager plugins """
+
+    #: A command to run to check whether the package manager is available on
+    #: a guest.
+    probe_command: Command
+
+    command: Command
+    options: Command
+
+    def __init__(self, *, guest: 'Guest', logger: tmt.log.Logger) -> None:
+        super().__init__(logger=logger)
+
+        self.guest = guest
+        self.command, self.options = self.prepare_command()
+
+    def prepare_command(self) -> tuple[Command, Command]:
+        """ Prepare installation command and subcommand options """
+        raise NotImplementedError
+
+    def _build_presence_script(self, *packages: Installable) -> ShellScript:
+        raise NotImplementedError
+
+    def check_presence(self, *packages: Package) -> CommandOutput:
+        return self.guest.execute(self._build_presence_script(*packages))
+
+    @overload
+    def install(
+            self,
+            *packages: Package,
```

Comment:
Are the `Package`/`Path` signatures supposed to be mutually exclusive with one another?

Reply:
Good question. They are not in real life, right? One can run `dnf install gcc /a/path/to/foo.rpm`, combining both. But in tmt code, they seemed to be never used at the same command. There was a path installing packages from `recommend`/`require`, and a path installing local RPMs. So, I followed the lead, but I think it's something we may change in the future.

Reply:
I'm not sure how `apt` does it, and I think `homebrew` does not support out-of-repo installs. But than again, these are just type-hints that would be used by the writers of the package manager plugin developers, so it's not crucial to be 100% truthful there.

Reply:
It turned out the answer is "it's complicated", and it depends on the package manager. `dnf` and `yum` are tolerant, one can mix package names with filesystem paths, package files and URLs, and everything works. `apt` is not so eager to help, and there it's just package + package file. So, these methods now accept all "installable classes", and have to deal with them. There are now some places where a package manager refuses to take action, e.g. our code will refuse to work with URLs (we would have to fetch & save into a temporary file, not bothered to do so now...).

Reply:
Seems to be covered by using `Installable`, resolving the conversation.